// angular-dual-listbox/esm2020/lib/angular-dual-listbox.module.mjs
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { FormsModule } from '@angular/forms';
import { DualListComponent } from './dual-list.component';
import * as i0 from "@angular/core";
export class AngularDualListBoxModule {
}
AngularDualListBoxModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "14.1.0", ngImport: i0, type: AngularDualListBoxModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
AngularDualListBoxModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "14.1.0", ngImport: i0, type: AngularDualListBoxModule, declarations: [DualListComponent], imports: [CommonModule,
        FormsModule], exports: [DualListComponent] });
AngularDualListBoxModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "14.1.0", ngImport: i0, type: AngularDualListBoxModule, imports: [CommonModule,
        FormsModule] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "14.1.0", ngImport: i0, type: AngularDualListBoxModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [
                        CommonModule,
                        FormsModule
                    ],
                    declarations: [DualListComponent],
                    exports: [DualListComponent]
                }]
        }] });
