// @wfpena/angular-wysiwyg/esm2022/lib/angular-editor.module.mjs
import { NgModule } from '@angular/core';
import { AngularEditorComponent } from './angular-editor.component';
import { AngularEditorToolbarComponent } from './angular-editor-toolbar.component';
import { FormsModule, ReactiveFormsModule } from '@angular/forms';
import { CommonModule } from '@angular/common';
import { AeSelectComponent } from './ae-select/ae-select.component';
import { AeButtonComponent } from './ae-button/ae-button.component';
import { AeToolbarSetComponent } from './ae-toolbar-set/ae-toolbar-set.component';
import * as i0 from "@angular/core";
export class AngularEditorModule {
    static ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "17.0.8", ngImport: i0, type: AngularEditorModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
    static ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "17.0.8", ngImport: i0, type: AngularEditorModule, declarations: [AngularEditorComponent,
            AngularEditorToolbarComponent,
            AeSelectComponent,
            AeButtonComponent,
            AeToolbarSetComponent], imports: [CommonModule, FormsModule, ReactiveFormsModule], exports: [AngularEditorComponent,
            AngularEditorToolbarComponent,
            AeButtonComponent,
            AeToolbarSetComponent] });
    static ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "17.0.8", ngImport: i0, type: AngularEditorModule, imports: [CommonModule, FormsModule, ReactiveFormsModule] });
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "17.0.8", ngImport: i0, type: AngularEditorModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [CommonModule, FormsModule, ReactiveFormsModule],
                    declarations: [
                        AngularEditorComponent,
                        AngularEditorToolbarComponent,
                        AeSelectComponent,
                        AeButtonComponent,
                        AeToolbarSetComponent,
                    ],
                    exports: [
                        AngularEditorComponent,
                        AngularEditorToolbarComponent,
                        AeButtonComponent,
                        AeToolbarSetComponent,
                    ],
                }]
        }] });
